File: src/org/perses/tree/spar/AbstractTreeEditAction.java
/*
 * Copyright (C) 2018-2020 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */
package org.perses.tree.spar;

/** The base class for editing a tree. */
public abstract class AbstractTreeEditAction implements Comparable<AbstractTreeEditAction> {

  final AbstractSparTreeNode targetNode;

  protected AbstractTreeEditAction(AbstractSparTreeNode targetNode) {
    this.targetNode = targetNode;
  }

  public final AbstractSparTreeNode getTargetNode() {
    return targetNode;
  }

  public final NodeDeletionAction asNodeDeletionAction() {
    return (NodeDeletionAction) this;
  }

  public final ChildHoistingAction asChildHoistingAction() {
    return (ChildHoistingAction) this;
  }

  @Override
  public final int compareTo(AbstractTreeEditAction o) {
    return internalCompareTo(o);
  }

  protected abstract int internalCompareTo(AbstractTreeEditAction o);

  public abstract String getDescription();

  @Override
  public String toString() {
    return getDescription();
  }

  @Override
  public final boolean equals(Object obj) {
    if (obj == null) {
      return false;
    }
    if (obj == this) {
      return true;
    }
    if (getClass() != obj.getClass()) {
      return false;
    }
    final AbstractTreeEditAction other = (AbstractTreeEditAction) obj;
    if (targetNode != other.targetNode) {
      return false;
    }
    return specificEquals(other);
  }

  protected abstract boolean specificEquals(AbstractTreeEditAction other);

  @Override
  public final int hashCode() {
    return 31 * specificHashCode() + targetNode.getNodeId();
  }

  protected abstract int specificHashCode();
}

File: src/org/perses/spartree/AbstractTreeEditAction.kt
/*
 * Copyright (C) 2018-2026 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */
package org.perses.spartree

import com.google.common.base.MoreObjects

/** The base class for editing a tree.  */
sealed class AbstractTreeEditAction : Comparable<AbstractTreeEditAction> {
  override fun compareTo(other: AbstractTreeEditAction): Int {
    val classCmp =
      compareBy<AbstractTreeEditAction> {
        it::class.java.canonicalName
      }.compare(this, other)
    return if (classCmp != 0) {
      classCmp
    } else {
      check(this::class.java == other::class.java)
      internalCompareTo(other)
    }
  }

  protected abstract fun internalCompareTo(o: AbstractTreeEditAction): Int

  abstract val conciseDescription: String

  final override fun toString(): String =
    MoreObjects.toStringHelper(this).addValue(conciseDescription).toString()

  abstract override fun equals(other: Any?): Boolean

  abstract override fun hashCode(): Int

  abstract fun apply()
}

abstract class AbstractTargetedTreeEditAction(
  val targetNode: AbstractSparTreeNode,
) : AbstractTreeEditAction() {
  private var used = false

  override fun apply() {
    check(!used)
    used = true
    internalApply()
  }

  abstract fun internalApply()

  override fun equals(other: Any?): Boolean {
    if (other == null) {
      return false
    }
    if (other === this) {
      return true
    }
    if (javaClass != other.javaClass) {
      return false
    }
    val otherAction = other as AbstractTargetedTreeEditAction
    return if (targetNode !== otherAction.targetNode) {
      false
    } else {
      specificEquals(otherAction)
    }
  }

  protected abstract fun specificEquals(other: AbstractTargetedTreeEditAction): Boolean

  override fun hashCode(): Int =
    31 * specificHashCode() + targetNode.nodeId + javaClass.hashCode() * 31 * 31

  protected abstract fun specificHashCode(): Int
}
